File: src/test/java/learnyouakotlin/part1/SessionTests.java
package learnyouakotlin.part1;

import org.junit.jupiter.api.Test;

import static java.util.Arrays.asList;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class SessionTests {

    Session original = new Session("The Title", null, new Slots(1, 2), new Presenter("Alice"));

    @Test
    public void can_change_presenters() {
        assertEquals(
            new Session("The Title", null, new Slots(1, 2), new Presenter("Bob"), new Presenter("Carol")),
            original.withPresenters(asList(new Presenter("Bob"), new Presenter("Carol"))));
    }

    @Test
    public void can_change_title() {
        assertEquals(
            new Session("Another Title", null, new Slots(1, 2), new Presenter("Alice")),
            original.withTitle("Another Title"));
    }

    @Test
    public void can_change_subtitle() {
        assertEquals(
            new Session("The Title", "The Subtitle", new Slots(1, 2), new Presenter("Alice")),
            original.withSubtitle("The Subtitle"));
    }

    @Test
    public void can_remove_subtitle() {
        assertEquals(
            new Session("The Title", null, new Slots(1, 2), new Presenter("Alice")),
            original.withSubtitle("The Subtitle").withSubtitle(null));
    }
}

File: src/test/java/learnyouakotlin/part1/SessionTests.kt
package learnyouakotlin.part1

import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert
import org.junit.Test

class SessionTests {
    var original = Session("The Title", null, Slots(1, 2), Presenter("Alice"))
    @Test
    fun can_change_presenters() {
        MatcherAssert.assertThat(
            original.withPresenters(listOf(Presenter("Bob"), Presenter("Carol"))),
            CoreMatchers.equalTo(Session("The Title", null, Slots(1, 2), Presenter("Bob"), Presenter("Carol")))
        )
    }
    
    @Test
    fun can_change_title() {
        MatcherAssert.assertThat(
            original.withTitle("Another Title"),
            CoreMatchers.equalTo(Session("Another Title", null, Slots(1, 2), Presenter("Alice")))
        )
    }
    
    @Test
    fun can_change_subtitle() {
        MatcherAssert.assertThat(
            original.withSubtitle("The Subtitle"),
            CoreMatchers.equalTo(Session("The Title", "The Subtitle", Slots(1, 2), Presenter("Alice")))
        )
    }
    
    @Test
    fun can_remove_subtitle() {
        MatcherAssert.assertThat(
            original.withSubtitle("The Subtitle").withSubtitle(null),
            CoreMatchers.equalTo(Session("The Title", null, Slots(1, 2), Presenter("Alice")))
        )
    }
}